src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/Build/OutputGroupProvider.cs

```diff
@@ -0,0 +1,95 @@
+        private readonly ConfiguredProject _configuredProject;
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="OutputGroupProvider"/> class.
+        /// </summary>
+        /// <param name="projectAccessor">Imported <see cref="IProjectAccessor"/>.</param>
+        /// <param name="configuredProject">Imported <see cref="ConfiguredProject"/>.</param>
+        [ImportingConstructor]
+        private OutputGroupProvider(IProjectAccessor projectAccessor, ConfiguredProject configuredProject)
+        {
+            _projectAccessor = projectAccessor;
+            _configuredProject = configuredProject;
+            _outputGroups = new AsyncLazy<IImmutableSet<IOutputGroup>>(GetOutputGroupMetadataAsync);
+        }
+
+        /// <inheritdoc/>
+        public Task<IImmutableSet<IOutputGroup>> OutputGroups
+        {
+            get { return _outputGroups.GetValueAsync(); }
+        }
+
+        /// <summary>
+        /// Gets a collection of names of targets in this project.
+        /// </summary>
+        /// <returns>Collection of the names of targets in this project.</returns>
+        private Task<ImmutableHashSet<string>> GetProjectTargetsAsync()
+        {
+            return _projectAccessor.OpenProjectForReadAsync(_configuredProject, project =>
+                project.Targets.Keys.ToImmutableHashSet(StringComparer.OrdinalIgnoreCase));
+        }
+
+        /// <summary>
+        /// Produces a set of output groups that is a subset of all well known output groups.
+        /// </summary>
+        /// <returns>Set of all known output groups.</returns>
+        private async Task<IImmutableSet<IOutputGroup>> GetOutputGroupMetadataAsync()
+        {
+            // Start with the comment set of output groups.
+            var result = s_outputGroups;
+
+            // Remove any well known output group for which no target is defined.
+            var targets = await GetProjectTargetsAsync();
+            foreach (var outputGroup in result)
+            {
+                if (!targets.Contains(outputGroup.TargetName))
+                {
+                    result = result.Remove(outputGroup);
+                }
+            }
+
+            return result;
+        }
+
+        /// <summary>
+        /// Creates a new output group with metadata only.
+        /// </summary>
+        /// <param name="name">Output group name.</param>
+        /// <param name="targetName">Output group target name.</param>
+        /// <param name="displayName">Output group display name.</param>
+        /// <param name="description">Optional output group description.</param>
+        /// <return>New <see cref="IOutputGroup"/>.</return>
+        private static IOutputGroup NewGroup(string name, string targetName, string displayName, string description)
+        {
+            return new OutputGroup(name, targetName, displayName, description, ImmutableList<KeyValuePair<string, IImmutableDictionary<string, string>>>.Empty, false);
```

Comment:
What does false for IsSuccesful mean?

Reply:
In CPS it's used to indicate whether or not the output group built successfully.  For our purposes it's not used at all.